Tests/SourceKitLSPTests/SemanticTokensTests.swift

```diff
@@ -0,0 +1,252 @@
+  var sk: TestClient! = nil
+
+  override func tearDown() {
+    sk = nil
+    connection = nil
+  }
+
+  override func setUp() {
+    connection = TestSourceKitServer()
+    sk = connection.client
+    _ = try! sk.sendSync(InitializeRequest(
+      processId: nil,
+      rootPath: nil,
+      rootURI: nil,
+      initializationOptions: nil,
+      capabilities: ClientCapabilities(
+        workspace: .init(
+          semanticTokens: .init(
+            refreshSupport: true
+          )
+        ),
+        textDocument: nil
+      ),
+      trace: .off,
+      workspaceFolders: nil
+    ))
+  }
+
+  private func performSemanticTokensRequest(text: String, range: Range<Position>? = nil) -> [SemanticToken] {
+    let url = URL(fileURLWithPath: "/\(#function)/a.swift")
+
+    // We wait for the first refresh request to make sure that the semantic tokens are ready
+
+    let expectation = XCTestExpectation(description: "performSemanticTokensRequest - refresh received")
+    sk.handleNextRequest { (req: Request<WorkspaceSemanticTokensRefreshRequest>) in
+      expectation.fulfill()
+      req.reply(VoidResponse())
+    }
+
+    sk.send(DidOpenTextDocumentNotification(textDocument: TextDocumentItem(
+      uri: DocumentURI(url),
+      language: .swift,
+      version: 17,
+      text: text
+    )))
+
+    let waitResult = XCTWaiter.wait(for: [expectation], timeout: 15)
+    if waitResult != .completed {
+      fatalError("error \(waitResult) while waiting for semantic token refresh")
+    }
+
+    let textDocument = TextDocumentIdentifier(url)
+    let response: DocumentSemanticTokensResponse!
+
+    if let range = range {
+      response = try! sk.sendSync(DocumentSemanticTokensRangeRequest(textDocument: textDocument, range: range))
+    } else {
+      response = try! sk.sendSync(DocumentSemanticTokensRequest(textDocument: textDocument))
+    }
+
+    return decodeFromIntArray(rawSemanticTokens: response.data)
+  }
+
+  func testEmpty() {
+    let text = ""
+    let tokens = performSemanticTokensRequest(text: text)
+    XCTAssertEqual(tokens, [])
+  }
+
+  func testRanged() {
+    let text = """
+    let x = 1
+    let y = 2
+    let z = 3
+    let w = 4
+    """
+    let start = Position(line: 1, utf16index: 0)
+    let end = Position(line: 2, utf16index: 5)
+    let tokens = performSemanticTokensRequest(text: text, range: start..<end)
+    XCTAssertEqual(tokens, [
+      SemanticToken(start: Position(line: 1, utf16index: 0), length: 3, kind: .keyword),
+      SemanticToken(start: Position(line: 1, utf16index: 4), length: 1, kind: .variable, modifiers: .declaration),
+      SemanticToken(start: Position(line: 1, utf16index: 8), length: 1, kind: .number),
+      SemanticToken(start: Position(line: 2, utf16index: 0), length: 3, kind: .keyword),
+      SemanticToken(start: Position(line: 2, utf16index: 4), length: 1, kind: .variable, modifiers: .declaration),
+    ])
+  }
+
+  func testIntArrayCoding() {
```

Comment:
Worth moving this case up to the test case as it’s slightly lower level then the rest, which are operating on source files?